src/proxy/pipeline/upstream.rs

```diff
@@ -1,232 +1,295 @@
+    pub(crate) log: slog::Logger,
+    pub(crate) metrics: Metrics,
+    pub(crate) downstream_socket: Arc<UdpSocket>,
+    pub(crate) filter_manager: SharedFilterManager,
 }
 
-impl Packet {
-    pub fn new(dest: SocketAddr, contents: Vec<u8>) -> Packet {
-        Packet { dest, contents }
-    }
+impl UpstreamDistributor {
+    pub fn spawn(
+        self,
+        mut shutdown_rx: ShutdownRx,
+    ) -> (DistributorTx, JoinHandle<Result<()>>) {
+        let manager: dashmap::DashMap<(SocketAddr, SocketAddr), UpstreamTx> =
+            dashmap::DashMap::new();
+        let (distributor_tx, mut distributor_rx) = distributor_queue();
+
+        let handle = tokio::spawn(async move {
+            loop {
+                debug!(self.log, "Awaiting incoming packet");
+                tokio::select! {
+                    received = distributor_rx.recv() => {
+                        match received {
+                            Some((from, endpoint, contents)) => {
+                                self.process(&manager, &shutdown_rx, from, endpoint, contents).await?;
+                            }
+                            None => {
+                                debug!(self.log, "Unable to retrieve message from downstream.");
+                                return Ok(())
+                            }
+                        }
+                    }
+                    _ = shutdown_rx.changed() => {
+                        debug!(self.log, "Closing Upstream");
+                        return Ok(());
+                    }
+                };
+            }
+        });
 
-    pub fn dest(&self) -> SocketAddr {
-        self.dest
+        (distributor_tx, handle)
     }
 
-    pub fn contents(&self) -> &Vec<u8> {
-        &self.contents
+    async fn process(
+        &self,
+        manager: &UpstreamMap,
+        shutdown_rx: &ShutdownRx,
+        from: SocketAddr,
+        endpoint: Endpoint,
+        contents: Vec<u8>,
+    ) -> Result<()> {
+        use dashmap::mapref::entry::Entry;
+        let new_upstream = || {
+            Upstream::new(
+                self.log.clone(),
+                self.filter_manager.clone(),
+                self.metrics.clone(),
+                self.downstream_socket.clone(),
+                from,
+                endpoint.clone(),
+            )
+        };
+
+        match manager.entry((from, endpoint.address)) {
+            Entry::Vacant(entry) => {
+                let sender = entry.insert((new_upstream)().await?.spawn(shutdown_rx.clone()));
+
+                sender.send(contents).await.map_err(Error::Sender)?;
+            }
+            Entry::Occupied(mut entry) => {
+                // If we get an error, then the session has expired, so
+                // insert a new upstream.
+                if let Err(_) = entry.get().send(contents).await {
+                    entry.insert((new_upstream)().await?.spawn(shutdown_rx.clone()));
+                }
+            }
```

Comment:
this seems more like an unwanted edge case rather than intended behavior which would be that the item id removed from the map when the session is no longer in use. it doesn't seem that we delete items from the map at all?

Reply:
Good catch, I'll add that depending on whether or not a LRU abstraction works out.